lib/client.c

```diff
@@ -1365,31 +1480,64 @@ SF_STATUS STDCALL _snowflake_execute_ex(SF_STMT *sfstmt,
     sfstmt->sequence_counter = ++sfstmt->connection->sequence_counter;
     _mutex_unlock(&sfstmt->connection->mutex_sequence_counter);
 
-    // TODO Do error handing and checking and stuff
-    ARRAY_LIST *p = (ARRAY_LIST *) sfstmt->params;
-    if (p && p->used > 0) {
+    if (_sf_get_current_param_style(sfstmt) == POSITIONAL)
+    {
         bindings = snowflake_cJSON_CreateObject();
-        for (i = 0; i < p->used; i++) {
+        for (i = 0; i < sfstmt->idx_len; i++)
+        {
             cJSON *binding;
-            input = (SF_BIND_INPUT *) sf_array_list_get(p, i + 1);
+            input = (SF_BIND_INPUT *) sf_param_store_get(sfstmt->params,
+                    i+1,NULL);
             if (input == NULL) {
                 continue;
             }
             // TODO check if input is null and either set error or write msg to log
             type = snowflake_type_to_string(
-                c_type_to_snowflake(input->c_type, SF_DB_TYPE_TIMESTAMP_NTZ));
+                    c_type_to_snowflake(input->c_type, SF_DB_TYPE_TIMESTAMP_NTZ));
             value = value_to_string(input->value, input->len, input->c_type);
             binding = snowflake_cJSON_CreateObject();
             char idxbuf[20];
             sprintf(idxbuf, "%lu", (unsigned long) (i + 1));
-          snowflake_cJSON_AddStringToObject(binding, "type", type);
-          snowflake_cJSON_AddStringToObject(binding, "value", value);
-          snowflake_cJSON_AddItemToObject(bindings, idxbuf, binding);
+            snowflake_cJSON_AddStringToObject(binding, "type", type);
+            snowflake_cJSON_AddStringToObject(binding, "value", value);
+            snowflake_cJSON_AddItemToObject(bindings, idxbuf, binding);
             if (value) {
                 SF_FREE(value);
             }
         }
     }
+    else if (_sf_get_current_param_style(sfstmt) == NAMED)
+    {
+        bindings = snowflake_cJSON_CreateObject();
+        for(i = 0; i < sfstmt->names_len; i++)
+        {
+            cJSON *binding;
+            input = (SF_BIND_INPUT *) sf_param_store_get(sfstmt->params,
+                    0,(char *)sfstmt->name_list[i]);
+            if (input == NULL)
+            {
+                continue;
+            }
+            type = snowflake_type_to_string(
+                    c_type_to_snowflake(input->c_type, SF_DB_TYPE_TIMESTAMP_NTZ));
+            value = value_to_string(input->value, input->len, input->c_type);
+            binding = snowflake_cJSON_CreateObject();
+
+            /*
+             * TODO - This may not be correct assumption for
+             * max name size
+             */
+            char name[MAX_NAME_SIZE];
```

Comment:
You can move this out of the for loop and dynamically allocate this buffer to something small (maybe 20 bytes or so). Then you can grow this buffer as needed. Since the `snowflake_CJSON_AddStringToObject` function just copies the string passed to it, you don't have to worry about persisting these values long term.

Reply:
This buffer was just a temporary buffer used to store the names. I realized this is not needed however, just pass (char *)(sfstmt->name_list[i]) directly to AddItemToObject.